micro-rdk/src/common/webrtc/signaling_server.rs

```diff
@@ -0,0 +1,176 @@
+            sender,
+            pending: Mutex::new(HashMap::new()),
+        }
+    }
+
+    pub fn optional_webrtc_config(
+        &self,
+        _request: OptionalWebRtcConfigRequest,
+    ) -> Result<OptionalWebRtcConfigResponse, ServerError> {
+        Ok(OptionalWebRtcConfigResponse { config: None })
+    }
+
+    pub async fn call(
+        &self,
+        request: CallRequest,
+        responses: Sender<Result<CallResponse, ServerError>>,
+    ) -> Result<(), ServerError> {
+        if request.disable_trickle {
+            return Err(ServerError::new(GrpcError::RpcUnimplemented, None));
+        }
+
+        // TODO: Can we re-use this engine?
+        let engine: Box<general_purpose::GeneralPurpose> = general_purpose::STANDARD.into();
+        let sdp_decoded = engine.decode(request.sdp.as_str()).map_err(|e| {
+            ServerError::new(
+                GrpcError::RpcInvalidArgument,
+                Some(Box::new(WebRtcError::InvalidSDPOffer(e.to_string()))),
+            )
+        })?;
+        let sdp_decoded: SdpOffer =
+            serde_json::from_slice(sdp_decoded.as_slice()).map_err(|e| {
+                ServerError::new(
+                    GrpcError::RpcInvalidArgument,
+                    Some(Box::new(WebRtcError::InvalidSDPOffer(e.to_string()))),
+                )
+            })?;
+        if sdp_decoded.sdp_type != "offer" {
+            return Err(ServerError::new(
+                GrpcError::RpcInvalidArgument,
+                Some(Box::new(WebRtcError::InvalidSDPOffer(format!(
+                    "unexpected type {}",
+                    sdp_decoded.sdp_type
+                )))),
+            ));
+        }
+        let mut cursor = Cursor::new(sdp_decoded.sdp);
+        let sdp = sdp::SessionDescription::unmarshal(&mut cursor).map_err(|e| {
+            ServerError::new(
+                GrpcError::RpcInvalidArgument,
+                Some(Box::new(WebRtcError::InvalidSDPOffer(e.to_string()))),
+            )
+        })?;
+        let sdp = Box::new(WebRtcSdp::new(sdp, Uuid::new_v4().into()));
+
+        // TODO: Should the channels be bounded?
+        let (update_tx, update_rx) = async_channel::bounded::<CallUpdateRequest>(1);
+        let (response_tx, response_rx) =
+            async_channel::bounded::<Result<CallResponse, ServerError>>(1);
+
+        let local_signaling = LocalSignaling {
+            tx: response_tx,
+            rx: update_rx,
+        };
+
+        let channel = Box::new(WebRtcSignalingChannel::new(
+            Either::Right(local_signaling),
+            sdp,
+        ));
+
+        match self.sender.send(channel).await {
+            Ok(()) => {
+                let uuid: Option<String> = None;
+                let mut uuid_guard = scopeguard::guard(uuid, |uuid| {
+                    if let Some(uuid) = uuid {
+                        self.pending.lock().unwrap().remove(uuid.as_str());
+                    }
+                });
+
+                loop {
```

Comment:
Here, we pull responses out of the channel as the api.rs machinery feeds them to us, and we stream them back out to the Sender that was passed in, which in turn results in a stream body up in grpc.rs.